beacon-chain/state/stategen/replay.go

```diff
@@ -308,18 +308,36 @@ func (s *State) genesisRoot(ctx context.Context) ([32]byte, error) {
 	return stateutil.BlockRoot(b.Block)
 }
 
-// This retrieves the archived root in the DB.
-func (s *State) archivedRoot(ctx context.Context, slot uint64) [32]byte {
+// This retrieves the highest archived root that represents the input slot in the DB.
+func (s *State) archivedRoot(ctx context.Context, slot uint64) ([32]byte, error) {
 	archivedIndex := uint64(0)
 	if slot/params.BeaconConfig().SlotsPerArchivedPoint > 1 {
 		archivedIndex = slot/params.BeaconConfig().SlotsPerArchivedPoint - 1
 	}
-	return s.beaconDB.ArchivedPointRoot(ctx, archivedIndex)
+
+	for archivedIndex > 0 {
+		if ctx.Err() != nil {
+			return [32]byte{}, ctx.Err()
+		}
+		if s.beaconDB.HasArchivedPoint(ctx, archivedIndex) {
+			return s.beaconDB.ArchivedPointRoot(ctx, archivedIndex), nil
+		}
+		archivedIndex--
```

Comment:
What is this? Can you add more to the godoc description about what this implementation is doing?

Reply:
will do